util/passport.js

```diff
@@ -100,6 +103,65 @@ function associateGoogle(user, accessToken, refreshToken, profile, done) {
     }).nodeify(done);
 }
 
+function authenticateGithub(accessToken, refreshToken, profile, done) {
+    db.withTransaction((dbClient) => {
+        return model.getByGithubAccount(dbClient, profile.id).then((rows) => {
+            if (rows.length > 0)
+                return model.recordLogin(dbClient, rows[0].id).then(() => rows[0]);
+
+            var username = profile.username || profile.email;
+            return model.create(dbClient, { username: username,
+                                            email: profile.email,
+                                            // we assume the email associated with a Github account is valid
+                                            // and we don't need extra validation
+                                            email_verified: true,
+                                            locale: 'en-US',
+                                            timezone: 'America/Los_Angeles',
+                                            github_id: profile.id,
+                                            human_name: profile.displayName,
+                                            cloud_id: makeRandom(8),
+                                            auth_token: makeRandom(),
+                                            storage_key: makeRandom() }).then((user) => {
+                user.newly_created = true;
+                return user;
+            });
+        });
+    }).then((user) => {
+        if (!user.newly_created)
+            return user;
+
+        // NOTE: we must start the user here because if we do it earlier we're
+        // still inside the transaction, and the master process (which uses a different
+        // database connection) will not see the new user in the database
+        return EngineManager.get().startUser(user.id).then(() => {
+            // asynchronously inject google-account device
+            EngineManager.get().getEngine(user.id).then((engine) => {
+                return engine.devices.loadOneDevice({ kind: 'com.github',
+                                                      profileId: profile.id,
+                                                      accessToken: accessToken,
+                                                      refreshToken: refreshToken }, true);
```

Comment:
This is not quite correct.
The goal of this piece of code is to automatically load the Github device in Almond if you login with Github (similar to how we load the Google skill if you login with Google). To do so, we must pass whatever parameters the Github device expects. The device is defined [here](https://github.com/stanford-oval/thingpedia-common-devices/blob/master/com.github/device.js), and you can find its OAuth code if you scroll down to `runOAuth2`